apps/yggdrasil/lib/yggdrasil/user.ex

```diff
@@ -29,14 +36,133 @@ defmodule Yggdrasil.User do
   end
 
   @doc """
-  adds password at end of chain protects the hashpwsalt from seeing a nil value
-  in the case of missing password field.
+  Creates the user and assigns the default roles of "player"
+  """
+  def create_with_default_role(attributes) do
+    changeset = User.create_changeset(%User{}, attributes)
+
+    with {:ok, user} <- Repo.insert(changeset),
+          :ok        <- assign_role(user, "player"),
+     do: {:ok, user}
+  end
+
+  def assign_role(user, role_name) do
+    role = Repo.get_by!(Role, name: role_name)
+
+    user_role = UserRole.changeset(%UserRole{},
+                                   %{user_id: user.id, role_id: role.id})
+    case Repo.insert(user_role) do
+      {:ok, _role} -> :ok
+      error        -> error
+    end
+  end
+
+  def load_permissions(user) do
+    query = from ur in UserRole,
+            where: ur.user_id ==  ^user.id,
+            preload: [role: [:role_permissions]]
+
+    permissions = query
+    |> Repo.all()
+    |> Enum.flat_map(fn ur -> ur.role.role_permissions end)
+    |> Enum.group_by(fn rp -> rp.resource end)
+    |> Enum.map(fn {r, rps} -> {String.to_atom(r), Enum.map(rps, fn rp -> String.to_atom(rp.permission) end)} end)
+    |> Enum.into(%{})
+
+    %{ user | permissions: permissions}
+  end
+
+  @doc ~S"""
+  Checks the list of resource/permissions pair given against what the user
+  has been granted and returns true if and only if all permissions supplied
+  have been granted for the user.
+
+  ## roles
+  A user can have one or more roles with each role having a set of resources
+  and a set of resources and those resources having a set of permissions. Roles
+  are not checked directly and only serve as a grouping mechanism for resources.
+
+  Since the user can have multiple roles, the permission set checked for a given
+  resource is the unique set across all the roles for that resource.
+
+  So if the user has `Role1` and `Role2` defined as follows:
+
+  * Role1
+    * foo
+      * :read
+      * :write
+  * Role2
+    * foo
+      * :read
+      * :all
+
+  The set of permissions checked for foo will be `[:read, :write, :all]`
+
+  ## Examples
+      # checking a single resource
+      User.is_granted? user, foo: [:read]
+
+      # checking multiple resources is allowed as well
+      User.is_granted? user, foo: [:read], bar: [:read, :write]
   """
-  def hash_password(changeset = %{:valid? => false}) do
+  def is_granted?(_user, []) do
+    {:error, "res_perm list was empty"}
+  end
+
+  def is_granted?(user, res_perms) do
```

Comment:
I think the imperative way the code flows was kind of getting to me, so I thought I would take a crack at an approach. This is completely untested but would love to have your feedback and see if it helps.

```elixir
def is_granted?(user, res_perms) do
    res_perms
        |> Enum.map(fn({res, perms} -> {res, user.permissions[res], perms})
        |> Enum.map(&has_permissions/1)
        |> Enum.filter(fn({ :error , _ }) -> true end)
        # if no errors were found the default accumulator will be returned, `:ok`
        |> Enum.reduce(:ok, fn(
            { :error, res_error }, :ok -> { :error, [res_error] }
            { :error, { res, meta }, { :error, errors } -> 
                # each resource should only have a single error associated to it from has_permissions/1
                { :error, Keyword.put(errors, res, meta) }
            end)
end

defp has_permissions({res, nil, _}), do: { :error, { res, [ :no_access, [] ] } }
defp has_permissions({res, _, []}), do: { :ok, res }
defp has_permissions({res, given, required) do
    given = MapSet.new given
    required = MapSet.new required

    cond MapSet.subset?(given, required) do
        true -> { :ok, res }
        false -> { :error, { res, [ :missing, MapSet.difference(required, given) ] } }
    end
end
```